lib/puppet/state_machine/machine_builder.rb

```diff
@@ -0,0 +1,182 @@
+require 'puppet/state_machine/state'
+
+module Puppet
+  class StateMachine
+    # The MachineBuilder class provides a DSL for implementing the states and transitions of a state machine,
+    # and validating that the machine is well behaved.
+    class MachineBuilder
+      # @param machine_name [String] The name of the machine to build.
+      def initialize(machine_name)
+        @machine_name = machine_name
+        @states = {}
+        @start_state = nil
+      end
+
+      # Build and verify a new state machine.
+      # @return [Puppet::StateMachine]
+      def build(&blk)
+        yield self
+        verify!
+        Puppet::StateMachine.new(@machine_name, @states, @start_state)
+      end
+
+      # Compose multiple state machines into a a single state machine.
+      #
+      # @example
+      #   machine.compose({machine1: m1, machine2: m2}) do |m|
+      #     m.start_state([:m1, :start])
+      #
+      #     m.transition(
+      #       source: [:m1, :final],
+      #       target: [:m2, :start],
+      #       on: Puppet::StateMachine::FINAL_EVENT)
+      #
+      #     m.final([:m2, :final])
+      #   end
+      def compose(machines, &blk)
+        machines.each_pair do |ns, machine|
+          @states.merge!(machine.namespaced_states(ns))
+        end
+        build(&blk)
+      end
+
+      # Verify that the machine is well defined.
+      # @return [void]
+      def verify!
+        verify_start_state!
+        verify_transition_targets!
+        verify_exiting_transitions!
+      end
+
+      # Define the start state for this state machine.
+      # @param name [Symbol]
+      # @return [void]
+      def start_state(name)
+        @start_state = name
+      end
+
+      # Generate a new state.
+      #
+      # @example
+      #   machine.state(:create_lockfile,
+      #     action: -> { Puppet::Util::Lockfile.new("/var/lock/puppet.lock").lock },
+      #     event: ->(has_lock) { has_lock ? :got_lock : :already_locked },
+      #     transitions: {
+      #       got_lock: :run_agent,
+      #       already_locked: :cancel_run,
+      #     })
+      #
+      # @return [Puppet::StateMachine::State] The generated machine state
+      def state(state_name, opts = {})
+        unhandled = opts.keys - [:action, :event, :type, :transitions]
```

Comment:
My reply to this apparently got eaten because I didn't submit the review properly. In any case, I would use named arguments if we could use them but since we need to support Ruby 1.9 syntax for JRuby we can't. The State class uses positional parameters, which is sensible for that class but less understandable for this DSL. The builder class glues together the named argument via hashes, and then translates that into the more compact positional arguments.

As for why the validation occurs here as opposed to within the state, the individual classes assume that they're being used correctly and defer all validation to this machine. The MachineBuilder class builds up a state machine incrementally instead of requiring all transitions and states to be defined in a single shot, and validates everything before the machine is generated. In short this just implements the builder pattern.

In addition, a single state only contains the context associated with that single state and is not aware of the entire machine. In order to do this validation within the state class we'd need to add a connection to the machine. In addition the state machine class is supposed to be immutable; pushing this logic into the State so that it can peer into the state machine would require that the state machine would have be modifiable after creation. Instead, the MachineBuilder class is the mutable object that incrementally builds up things, validates them, and then generates an immutable, verified state machine.

Another benefit of extracting all of the validation logic here is that it allows us to turn it off if desired. Once a state machine has been implemented we don't necessarily need to validate it every time; these validations only really need to occur while the implementation of the states are in flux.